### DISCUSSION TITLE
pyb & machine with I2C and STM32

### DISCUSSION THREAD

**Discussion by AutoPilotjn** [micropython/micropython#10612] [Hardware & Peripherals]
Suggestion.
I could not succeed in having pyb.i2c work with sending or receiving more than one byte. Indeed mem_read() did not return the proper bytes read, according to my trials.
Then I lately noticed that pyB.I2C shall be used for legacy I2C support only. Instead machine.i2c shall be used.
Maybe I overlooked at the documentaion, maybe I did not manage to make pyb mem_read & mem_write to work ?
In any case, may I suggest to make it clearer in the documentation that 'machine' shall be used ?
For the sake of clarity machine.i2c.writeto_mem() & readfrom_mem() work satisfactorily as described in the documentation.

**Comment by jimmo** [comment]
> In any case, may I suggest to make it clearer in the documentation that 'machine' shall be used ?

Yes, this is the TODO list.

> For the sake of clarity machine.i2c.writeto_mem() & readfrom_mem() work satisfactorily as described in the documentation.

OK that is good to hear. It is surprising that `pyb.I2C` was not working though as this is well-tested and has been in use for nearly 10 years. One thing to note is that `machine.I2C` supports "soft I2C" on any pin, and it's possible you're using it in that mode, whereas pyb.I2C was configured for the wrong pins.
